include/vcpkg/base/message_sinks.h

```diff
@@ -75,4 +77,30 @@ namespace vcpkg
         CombiningSink(MessageSink& first, MessageSink& second) : m_first(first), m_second(second) { }
         void print(Color c, StringView sv) override;
     };
+
+    struct BGMessageSink final : MessageSink
+    {
+        BGMessageSink(MessageSink& out_sink) : out_sink(out_sink) { }
+        ~BGMessageSink() { publish_directly_to_out_sink(); }
+        // must be called from producer
+        void print(Color c, StringView sv) override;
+        using MessageSink::print;
+
+        // must be called from consumer (synchronizer of out)
+        void print_published();
+
+        void publish_directly_to_out_sink();
+
+    private:
+        MessageSink& out_sink;
+
+        std::mutex m_published_lock;
+        std::vector<std::pair<Color, std::string>> m_published;
+
+        // buffers messages until newline is reached
+        // guarded by m_print_directly_lock
+        std::vector<std::pair<Color, std::string>> m_unpublished;
```

Comment:
@BillyONeal Now that you have implemented the "error document" type stuff, I am right that this should be a `DiagnosticLine` etc. instead.

Reply:
I'm not sure yet, I ran into a problem where I'm not sure exactly how "status" type information should be conveyed through this infrastructure; I'm working on it over here: https://github.com/microsoft/vcpkg-tool/compare/main...BillyONeal:vcpkg-tool:message-sink-line

As part of that I realized that https://github.com/microsoft/vcpkg-tool/pull/1137 touches the same area and would be easier to merge first.

Reply:
Why does a "status" need to be conveyed here at all? This are only informational messages. 
Or what is the problem you are trying to solve here?